### ISSUE TITLE
[feature request] verify checksum after upload

### ISSUE THREAD

**Issue by kcwu** [odeke-em/drive#115]
After upload, query checksum from google drive and compare to the checksum of what just uploaded.

**Comment by odeke-em** [comment]
This would have to be an optional mode.

**Comment by arniotis** [comment]
I have the same need - my solution was to create a new command "drive md5sum" which fetches md5sum from the server in a format compatible with the md5sum command. This way I can do

````
drive md5sum | md5sum -c
````
or, even
````
drive md5sum -r | md5sum -c
````
which fetches (very quickly) the md5sums from Drive and compares them to the md5sums on the local drive. Of course this also works after a download, or with pre-computed md5sum output files

````
$ drive md5sum  >  local
$ md5sum *      >  remote
$ diff ~/remote ~/local
````

This could also be useful in comparing checksums for files across different Google Drive accounts

@odeke-em would you be interested in incorporating this into drive? Let me know if you are amenable to it. and I can do a pull request. 

Cheers,

Michael

**Comment by odeke-em** [comment]
Hello @arniotis welcome to drive and thank you very much for looking into this!
However for @kcwu's case they might just want to do it right after upload since they might be processing other files. drive has  already functionality for computing checksums of files so we could do this in code.

As for your feature, drive has `stat` that returns information about a file including checksum, I think I'd like this to be a subcommand of stat that just returns md5checksum.
Please let me know what you think, and thank you again.

**Comment by arniotis** [comment]
That was my first thought as well - in fact I used `stat` as a `sta(r)t`. The problem was that `stat`'s   output is inherently unordered as it is highly parallelized. 

`drive md5sum` needs to produce its output sorted by filename, so it can be used as input to `diff` (see my example of comparing across two Drive accounts, above). Yes, one can pipe through `sort -k 1.33` every time, but that over-complicates things for the end user (and piping through sort means the user has to wait for both scans to completely finish before the first results come up)

I looked at reworking stat to sort its output, and it would have been way too invasive and error-prone for me to do so. 

My changes are relatively small (see https://github.com/odeke-em/drive/compare/master...arniotis:master ) and `drive md5sum` runs about 12 times faster than `drive stat` for my files (since it has so much less to do...)

Let me know what you think.

Cheers, and many thanks for all the great work you are putting into this.

Michael

**Comment by arniotis** [comment]
Oh, and as for @kcwu's request: One simple way to satisfy it would be to allow for a new option in `drive push`, let's call it -echo, which simply echoes back each filename right after it is successfully uploaded. Then one could run

```shell
$ drive push -no-prompt -echo | xargs drive md5sum | md5sum -c 
```

Which would check each file right after upload. -echo could come handy for other uses. For example, this deletes each file right after upload:

```shell
$ drive push -no-prompt -no-clobber -echo | xargs rm 
```

**Comment by odeke-em** [comment]
@arniotis nice, that's cool!

Now in regards to functionality:
```
That was my first thought as well - in fact I used stat as a sta(r)t. The problem was that stat's output is inherently unordered as it is highly parallelized.
```
That can be fixed and made sequential -- in fact this is becoming sequential since with extra depth traversal `stat` produces garbage output. Actually another option would be to use `list` and this way the semantics will be the same and enable exact traversal etc.

``
let's call it -echo
``
I like that. In fact, some arg like that can be used to fulfill @kcwu's request.

As for the new code, I think that is an entirely new command inspired by this issue but it won't solve this issue. I think `stat` and `list` can be adapted to match the proposed `md5sum` functionality. 

Otherwise, please go ahead and make the PR and I'll give you a review. Would you mind creating an issue proposing this new functionality? (as a request to fill in the PR). This way in the next release I'll be able to track what's gone on and give you the credit :), as well as easily locate the origins for the feature in the future.
Good stuff @arniotis and thank you :+1:

**Comment by odeke-em** [comment]
btw @arniotis https://github.com/odeke-em/drive/commit/1c7e238a67bd20bb3a1122647d59125f8348051a has sequential stating.

**Comment by arniotis** [comment]
Yes... That's pretty much what I did too, took stat and removed all the threading to make it sequential...

Looking at this new version of stat, I can now easily port my code to it, so that the command is `drive stat -md5sum` instead of `drive md5sum`.  Would you prefer this?

**Comment by odeke-em** [comment]
Actually you can incorporate md5sum as a switch so a subcommand as well of `stat`. This way we shall have:
`drive stat --md5sum` == `drive md5sum`
Please take a look at the usage of:
`drive about --quota` == `drive quota`
OR
`drive about --features` == `drive features`
as you'll can see the sub commands are just spin offs and some sort of switches.
